feat: implement DIP0026

panleone opened the pull request:
Implement DIP0026 following the official documentation https://github.com/dashpay/dips/blob/master/dip-0026.md
With this PR it is possible to create a masternode that havs up to 32 payees and it is a first step toward trustless masternode shares 

## What was done?
In ProTxs the field `CScript scriptPayout` has been changed to `std::vector<PayoutShare> payoutShares;` which is a vector of scripts/ rewards shares.
Serialiation and  Deserialization of ProTxs has been changed accordingly
Finally I added the regtest parameters to activate DIP0026 with a soft fork 

## How Has This Been Tested?
 I have added unit tests that test creation, consensus and payment of masternodes with more than one payee

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [ ] I have assigned this pull request to a milestone

panleone committed:
feat(consensus): Implement DIP0026

Expand CProRegTx and CProUpRegTx according to DIP0026 standard, change serialization and deserialization methods in order to take in account the new field,
Add consensus rules specified in the DIP0026 (max 32 payees, sum of payment shares cannot be greater than 10000,...)
change dmnstate in order to take in account the new field
change consensus rules in such a way that masternodes will pay all the payees
add multiple payees in the qt masternode table

panleone committed:
feat(rpc): Update RPC

panleone committed:
test: Update existing functional test

knst commented on src/test/evo_trivialvalidation.cpp:
```diff
@@ -29,7 +29,7 @@ void TestTxHelper(const CMutableTransaction& tx, bool is_basic_bls, bool expecte
 
     TxValidationState dummy_state;
-    BOOST_CHECK_EQUAL(payload.IsTriviallyValid(is_basic_bls, dummy_state), !expected_failure);
+    BOOST_CHECK_EQUAL(payload.IsTriviallyValid(is_basic_bls, is_multi_payout_active, dummy_state), !expected_failure);
```
I think  the  list of `trivially_valid` and `trivially_invalid` should have at least one transaction of each type with multiple payees to be sure that serialization works correctly.

panleone replied:
moved most of the unit test coverage to `trivially_valid` `trivially_invalid`

knst reviewed (commented):
please check my comments for PR.

That's not a final version of my review, but that's what I have seen at the moment

panleone commented on src/evo/providertx.h:
```diff
@@ -203,45 +263,72 @@ class CProUpRegTx
 
-    SERIALIZE_METHODS(CProUpRegTx, obj)
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action)
```
I changed to SerializationOp because in this way I could differentiate between reading and writing to have backward compatibility when reading a message with an older version:

if deserializing with a message with an older  version message just read a CScript:
`

    CScript payoutScript;
    READWRITE(payoutScript);
    payoutShares = {PayoutShare(payoutScript)};        
`

if serializing a message with an older version write the first element of the array only:

`
     READWRITE(payoutShares[0].scriptPayout);
`

am I missing something?

knst replied:
hm, does [this way](https://github.com/knst/dash/commit/65ab6511515b3cf9ee598e694133a7520cfedfd4) do same thing or I am missing something?

knst replied:
We can have a default value for member rewardShare as 10.000 then

panleone commented on src/evo/providertx.h:
```diff
@@ -45,43 +75,64 @@ class CProRegTx
+            keyIDVoting,
+            nOperatorReward);
+        if (nVersion < MULTI_PAYOUT_VERSION) {
+            if (ser_action.ForRead()) {
```
the idea was just differentiating between serializing and deserializing  if the version is older  than `MULTI_PAYOUT_VERSION `(see the other comment)

knst commented on src/qt/masternodelist.cpp:
```diff
@@ -244,9 +249,14 @@ void MasternodeList::updateDIP3List()
+                payees = (i == 0) ? EncodeDestination(payeeDest) : (payees + ", " + EncodeDestination(payeeDest));
+            }
         }
+        payeeStr = QString::fromStdString(payees);
```
if ExtractDestionation will fail the size of array would be wrong.
Let's add instead for each failed "UNKNOWN"

something like:
```
QString payeeStr;
for (const auto& payee : dmn.pdmnState->payoutShares) {
   if (!payeeStr.empty()) payeeStr += ", ";
   if (ExtractDestination(payeeShare.scriptPayout, payeeDest)) {
     payees += QString::fromStdString(EncodeDestination(payeeDest));
  else payees += tr("UNKNOWN");
}
```

panleone commented on src/rpc/evo.cpp:
```diff
@@ -997,8 +1015,8 @@ static UniValue protx_update_service_common_wrapper(const JSONRPCRequest& reques
-            // use payout address as default source for fees
-            ExtractDestination(dmn->pdmnState->scriptPayout, feeSource);
+            // use first payout address as default source for fees
+            ExtractDestination(dmn->pdmnState->payoutShares[0].scriptPayout, feeSource);
```
Ok here the situation is a bit more tricky because the "source for fees" is the address that spends an `utxo` and receives back `utxo - (fees needed to fund the tx)` ... So if we want to replicate we would have to put all payees address also as inputs... and we would also check that all those addresses belongs to the wallet + each address has has at least 1 utxo to fund the tx.

At the same time I agree that the first address as nothing special compared to the others... so I suggest that we try to fund the tx with payees addresses only if there is only 1 payee,  what do you think?

panleone replied:
if there is more than 1 payee transaction can be still funded with operatorAddress or the address specified by the user in the rpc

knst commented:
please address this warnings from CI also:
```
The subject line of commit hash 6dbc490ff109af21ac2611b3b1f4fe4a7e207271 is followed by a non-empty line. Subject lines should always be followed by a blank line.
The subject line of commit hash 9a6f9e9f44d347646a9e47219a532a936ac99e73 is followed by a non-empty line. Subject lines should always be followed by a blank line.
src/evo/deterministicmns.cpp:1711:72: warning: Consider using std::any_of algorithm instead of a raw loop. [useStlAlgorithm]
src/evo/providertx.cpp:26:16: warning: Variable 'payoutShare' can be declared as reference to const [constVariable]
src/evo/providertx.cpp:68:16: warning: Variable 'payoutShare' can be declared as reference to const [constVariable]
src/rpc/evo.cpp:1319:71: warning: Consider using std::any_of algorithm instead of a raw loop. [useStlAlgorithm]
src/rpc/evo.cpp:1395:84: warning: Consider using std::any_of algorithm instead of a raw loop. [useStlAlgorithm]
```
https://gitlab.com/dashpay/dash/-/jobs/5871076186
they are cause CI tail.

panleone committed:
feat(consensus): Generalize regtest ehf

In this way any deployment that is activated with DIP0023 will be deployed on regtest after activating spork 24

panleone committed:
test: add activate_dip_0026

And improve dynamically_prepare_masternode in such a way to support multiple payees

panleone commented on test/functional/rpc_masternode.py:
```diff
@@ -74,13 +74,15 @@ def run_test(self):
             protx_info = self.nodes[0].protx("info", payments_masternode["proTxHash"])
             if len(payments_masternode["payees"]) == 1:
-                assert_equal(protx_info["state"]["payoutAddress"], payments_masternode["payees"][0]["address"])
+                assert_equal(protx_info["state"]["payouts"][0]["payoutAddress"], payments_masternode["payees"][0]["address"])
```
Added a functional test which covers only dip0026

PastaPastaPasta commented:
I haven't reviewed this yet; but thank you for your contribution!

panleone committed:
test: add dip0026 functional tests

panleone committed:
feat: add DIP0026 to regtest chain params

panleone committed:
test: DIP0026 unit tests

UdjinM6 commented on src/masternode/payments.cpp:
```diff
@@ -60,7 +60,12 @@
+            if (payeeReward > 0) {
+                voutMasternodePaymentsRet.emplace_back(payeeReward, payoutShare.scriptPayout);
+            }
+        }
```
There is a set of rules which really makes you want to keep Operator related things separated I think:
- only Operator must be able to change his payout script
- Operator must not be able to change anyone else's payouts

UdjinM6 commented:
Looks very solid. Needs some cleanups and small fixes but great job overall @panleone ! 👍

panleone commented on src/evo/dmnstate.h:
```diff
@@ -324,7 +330,7 @@ class CDeterministicMNStateDiff
     DMN_STATE_DIFF_LINE(keyIDVoting)                   \
     DMN_STATE_DIFF_LINE(addr)                          \
-    DMN_STATE_DIFF_LINE(scriptPayout)                  \
+    DMN_STATE_DIFF_LINE(payoutShares)                  \
```
Oh, nice catch! It is not that simple indeed since `CDeterministicMNStateDiff`  does not use the serialization of `CDeterministicMNState`, but I guess we can use the same trick we did with `pubKeyOperator`

panleone replied:
ok, now every `std::vector<PayoutShare>` is serialized with the same helper class (whose compatibility with older versions is tested in trivially_valid and trivially_invalid unit tests).

`CDeterministicMNStateDiff` also uses this class + from now on each time it serializes a vector of PayoutShare it serializes the version.

knst commented on src/evo/providertx.h:
```diff
@@ -21,16 +21,53 @@ class CBlockIndex;
     {
-        return is_basic_scheme_active ? BASIC_BLS_VERSION : LEGACY_BLS_VERSION;
+        if (is_multi_payout_active) {
+            // multi payout is activated after basic scheme
```
assert could cause app crash because. ptx.payouotShares is data that accepted from internet

```
ptx.nVersion = CProRegTx::GetVersion(!use_legacy, ptx.payoutShares.size() > 1);
```

UdjinM6 replied:
good point! 👍 pls ignore this suggestion then

UdjinM6 commented on src/evo/providertx.h:
```diff
@@ -203,18 +247,25 @@ class CProUpRegTx
     {
-        return is_basic_scheme_active ? BASIC_BLS_VERSION : LEGACY_BLS_VERSION;
+        if (is_multi_payout_active) {
+            // multi payout is activated after basic scheme
```
https://github.com/dashpay/dash/pull/5799#discussion_r1445274426

knst commented on src/llmq/ehf_signals.cpp:
```diff
@@ -54,8 +55,15 @@ void CEHFSignalsHandler::UpdatedBlockTip(const CBlockIndex* const pindexNew)
     }
-    // TODO: should do this for all not-yet-signied bits
-    trySignEHFSignal(Params().GetConsensus().vDeployments[Consensus::DEPLOYMENT_MN_RR].bit, pindexNew);
+
```
Unlikely we are going to have any other EHF fork before dip0026 is merged.
But to speed up process review, merging new code and reduce conflicts (dip0026 goes for v21, not v20.x) it's reasonable to split

panleone committed:
feat: Serialize and deserialize correctly

A helper class has been added which has the role to wrap and correctly serialize and deserialize a vector of payoutShare.

panleone committed:
[squash]: do not initialize payoutShares

panleone committed:
[squash]: revert changes to lines I did not touch

knst commented:
@panleone @PastaPastaPasta @UdjinM6 
I've just observed another crucial aspect. With the "mn_rr" hard fork, a portion of the masternodes' rewards will be reallocated between the masternodes directly and the platform reward (approximately 37%). The platform is responsible for managing its share of the reward and distributing it among evo nodes.

Therefore, once the mn_rr is activated (in conjunction with the platform release witch is expected before DIP0026 hard fork), we will require the platform to also support an array of payees.

panleone commented:
> @panleone @PastaPastaPasta @UdjinM6 I've just observed another crucial aspect. With the "mn_rr" hard fork, a portion of the masternodes' rewards will be reallocated between the masternodes directly and the platform reward (approximately 37%). The platform is responsible for managing its share of the reward and distributing it among evo nodes.
> 
> Therefore, once the mn_rr is activated (in conjunction with the platform release witch is expected before DIP0026 hard fork), we will require the platform to also support an array of payees.

Good point, then we must also implement dip0026 on platform. At the moment due to exams I don't have enough free time to do it, but since v21 is still far away it should not be a problem

panleone committed:
[squash]: use ranges in place of for loop

panleone committed:
[squash]: apply review

changes to isTriviallyValid and if formatting

panleone committed:
[squash]: include all payoutShares in ToString() logs

panleone committed:
[squash]: require fee address if tx is multi payee

knst commented on src/bloom.cpp:
```diff
@@ -112,14 +112,7 @@ bool CBloomFilter::CheckScript(const CScript &script) const
 
 bool CBloomFilter::CheckPayeeSharesScripts(const std::vector<PayoutShare>& payoutShares) const
```
nit: maybe better name is `cointainsPayeeShareScript` or something similar? `check` makes me confused like that's kind of validation, but ignore me if I am wrong so far as English is not my native lang

DashCoreAutoGuix commented:
## ℹ️ Backport Verification - Not Applicable

This PR implements DIP0026 (masternode payout shares), which is a **Dash-specific feature**, not a Bitcoin backport.

**Original Bitcoin commit:** N/A - This is a Dash-native implementation  
**Reviewed commit hash:** `4b5c4aaac1-verify-1753723627`

**Why verification was skipped:**
- This PR introduces new Dash functionality (DIP0026)
- No corresponding Bitcoin changes to verify against
- This is legitimate Dash development, not a backport

**Recommendation:** This PR should be reviewed through the standard Dash development process, not the backport verification system.

---
*Note: The backport verification system only applies to PRs that port changes from Bitcoin Core to Dash Core.*

thepastaclaw reviewed (changes requested):
<!-- thepastaclaw-review v1 sha=4b5c4aaac1dd8f0cea62871a4aac7045b732df95 -->
## Code Review

Found four blocking consensus/serialization issues in the DIP-0026 multi-payee implementation: (1) uint16_t accumulator in payout-share validation wraps, letting forged totals pass; (2) CDeterministicMNStateDiff deserializes payoutShares before nVersion, corrupting the stream for v3 diffs; (3) ProUpRegTx updates leave nVersion stale unless the operator key changes, silently truncating multi-payee data on subsequent serialization; (4) duplicate identical payout shares can underpay masternodes because the payment matcher uses set-membership semantics. Several lower-severity issues around input validation and migration fragility are also flagged.

*Reviewed commit: 4b5c4aaa*

🔴 4 blocking | 🟡 2 suggestion(s) | 💬 3 nitpick(s)

<details>
<summary><b>blocking:</b> uint16_t totalPayoutReward wraps; sum != 10000 can pass validation</summary>

`src/evo/providertx.cpp` (lines 15-41)

`totalPayoutReward` is `uint16_t` (max 65535), but with up to 32 payout shares each capped at 10000, the legitimate maximum sum is 320000. Arithmetic on `uint16_t` wraps modulo 65536, so an attacker can craft entries whose real sum is e.g. 75536 (= 65536 + 10000) — for instance, seven entries of 10000 plus one of 5536, each individually <= 10000 and the count <= 32 — and the accumulator settles at 10000, making `totalPayoutReward != 10000` evaluate false. The tx passes `TriviallyVerifyProRegPayees` even though the shares total >100%. Downstream, `GetBlockTxOuts` in src/masternode/payments.cpp:63-67 distributes `masternodeReward * payoutShareReward / 10000` per share, so the masternode would consume ~7.5x its share of the block reward and FillBlockPayments cannot produce a balanced coinbase — a chain-halt DoS vector that survives validation. Also note the order bug at line 32-35: the accumulator is incremented before the per-share `> 10000` check, so the bound check should happen first. Existing tests only exercise sum > 10000 without overflow, so the wrap path is uncovered.

</details>
<details>
<summary><b>blocking:</b> CDeterministicMNStateDiff reads payoutShares before nVersion, corrupting v3 diffs</summary>

`src/evo/dmnstate.h` (lines 314-379)

`DMN_STATE_DIFF_ALL_FIELDS` lists `payoutShares` (line 327) before `nVersion` (line 333). On SER_READ the macro iterates in declaration order: when `payoutShares` is processed, `obj.state.nVersion` is still the default `LEGACY_BLS_VERSION` (1), so `PayoutSharesSerializerWrapper(..., obj.state.nVersion < MULTI_PAYOUT_VERSION)` enters the single-payee branch and reads only one `CScript`. But if the diff was written with `nVersion == MULTI_PAYOUT_VERSION` (which the diff constructor at line 347 forces whenever `Field_payoutShares` is set), the stream actually encodes `vector<PayoutShare>` (count varint + repeated PayoutShare records each carrying an extra `uint16_t`). The wrong number of bytes are consumed and every field after — scriptOperatorPayout, nConsecutivePayments, platformNodeID, ports, finally nVersion itself — is decoded from the wrong offset. This produces silent state corruption as soon as any DB-loaded or wire-transmitted diff carries a DIP-0026 payout-share change. The `pubKeyOperator` precedent (with `SetLegacy(...)` after the fact) works only because BLS public keys are fixed-length across schemes; the same trick cannot fix payoutShares since the legacy and v3 encodings have different lengths. Fix: serialize `nVersion` first (write-format change) or special-case reading nVersion before payoutShares/pubKeyOperator. No unit test currently round-trips a CDeterministicMNStateDiff with multi-payout shares.

</details>
<details>
<summary><b>blocking:</b> ProUpRegTx payout-share update does not propagate state nVersion</summary>

`src/evo/deterministicmns.cpp` (lines 870-880)

`newState->nVersion` is only assigned inside the `if (newState->pubKeyOperator != proTx.pubKeyOperator)` branch. When a registrar update keeps the operator key but changes payoutShares (e.g. switches a single-payee MN to multi-payee, or vice versa), `newState->nVersion` remains the previous DMN state version while `newState->payoutShares` is overwritten with the new vector. The `CDeterministicMNState` serializer at src/evo/dmnstate.h:239 branches on `nVersion < MULTI_PAYOUT_VERSION`; for a stale version this takes the legacy single-payee path and silently drops every payee after `payoutShares[0]`. The corruption appears in-memory only after the next DB round-trip or state diff, so block validation accepts the ProUpRegTx but the resulting state is wrong. Set `newState->nVersion = proTx.nVersion;` unconditionally before applying the rest of the update.

</details>
<details>
<summary><b>blocking:</b> Duplicate payout shares can bypass masternode-payment enforcement</summary>

`src/evo/providertx.cpp` (lines 15-41)

`TriviallyVerifyProRegPayees` checks count, script type, per-share cap and total, but never rejects duplicate payout entries. A masternode can register two PayoutShares with identical scriptPayout and identical payoutShareReward. `GetBlockTxOuts` (src/masternode/payments.cpp:63) emits two identical `CTxOut`s, but the validator at src/masternode/payments.cpp:117-126 uses `ranges::any_of(txNew.vout, ...)` — set-membership semantics that do not consume matches. A malicious miner can include only one of the two identical outputs; both expected entries match the same actual output, the block passes validation, and the masternode is underpaid (the miner pockets the difference). This was not reachable pre-DIP-0026 because there was only one MN payee. Reject duplicates in `TriviallyVerifyProRegPayees` (e.g. dedup by (scriptPayout) — same script implies same destination — or fix the matcher to consume outputs as it pairs them).

</details>
<details>
<summary><b>suggestion:</b> RPC silently wraps payoutShareReward via int -> uint16_t conversion</summary>

`src/rpc/evo.cpp` (lines 549-573)

`payoutShares[i].get_array()[1].get_int()` returns a signed `int`, then is passed to `PayoutShare(CScript, uint16_t)` where the implicit narrowing wraps modulo 65536. Out-of-range inputs (e.g. -1, 70000) become 65535, 4464, and so on. The downstream `> 10000` validation in `TriviallyVerifyProRegPayees` catches some of these post-wrap, but the user sees an error referencing a value they never entered, and combined with the uint16_t accumulator overflow above, malformed RPC input can reach unintended states. Validate the int explicitly before constructing the PayoutShare.

</details>
<details>
<summary><b>suggestion:</b> Zero-value payoutShareReward is accepted</summary>

`src/evo/providertx.cpp` (lines 26-36)

DIP-0026 specifies share values in [1, 10000]; the validator only enforces the upper bound (`payoutShare.payoutShareReward > 10000`). A payee with share == 0 is accepted: it contributes nothing to the sum, produces no output in `GetBlockTxOuts` (which skips `payeeReward == 0`), but inflates the on-chain payee vector, wasting storage/bandwidth and creating a fingerprinting surface. The functional test helper `dynamically_prepare_masternode` (test/functional/test_framework/test_framework.py:1212) already uses `random.randint(0, ...)` for shares, so 0 is reachable through normal flows. Reject 0 as well.

</details>
<details>
<summary><b>nitpick:</b> PayoutShare default constructor leaves payoutShareReward = 0 while explicit ctor defaults to 10000</summary>

`src/evo/providertx.h` (lines 24-44)

`PayoutShare()` (the implicit default) uses the in-class initializer at line 28 and leaves `payoutShareReward = 0`. The explicit `PayoutShare(scriptPayout)` ctor defaults `payoutShareReward = 10000`. The legacy deserialization path (`PayoutSharesSerializerWrapper` at line 78) and the legacy-migration constructors in dmnstate.h:192,209 rely on the 10000 default to give the legacy sole payee 100% of the reward. New code that uses `PayoutShare{}` and forgets to set the share will silently produce 0 and trip `bad-protx-payee-reward-sum`. Make the two defaults consistent (drop the in-class `{0}` or have both ctors use the same canonical value).

</details>
<details>
<summary><b>nitpick:</b> Legacy DMNState migration constructs PayoutShare relying on implicit default reward</summary>

`src/evo/dmnstate.h` (lines 179-213)

Both migration constructors use `payoutShares({PayoutShare(s.scriptPayout)})`, which depends on the explicit ctor's default argument (10000) to assign 100% to the sole legacy payee. This is correct today but fragile: if the explicit ctor's default ever changes, every historical record loaded from the old DB formats would silently start paying its sole payee a different share. Pass the value explicitly: `PayoutShare(s.scriptPayout, 10000)`.

</details>
<details>
<summary><b>nitpick:</b> protx_revoke dereferences payoutShares[0] without a non-empty guard</summary>

`src/rpc/evo.cpp` (lines 1228-1238)

`dmn->pdmnState->payoutShares[0].scriptPayout != CScript()` indexes element 0 before any size check. In normal operation a registered MN must have at least one payout share, but adding a defensive `!payoutShares.empty()` predicate (mirroring the pattern used elsewhere) would be safer; the existing `if (payoutShares.size() > 1)` immediately below is unreachable if the vector were ever empty due to corruption.

</details>

_Inline posting via scripts/review_poster.py failed, so I posted the verified findings in the review body instead: Traceback (most recent call last):
  File "/Users/claw/.openclaw/workspace/scripts/review_poster.py", line 138, in <module>
    result = post_review(repo, pr_number, head_sha, verified, dry_run=dry_run)
  File "/Users/claw/.openclaw/workspace/scripts/review_poster.py", line 107, in post_review
    result = _gh(
  File "/Users/claw/.openclaw/workspace/scripts/review_poster.py", line 23, in _gh
    raise RuntimeError(f"gh {' '.join(args)} failed: {result.stderr.strip()}")
RuntimeError: gh api /repos/dashpay/dash/pulls/5799/reviews --method POST --input - failed: gh: Unprocessable Entity (HTTP 422)_
